### PR TITLE
Getting templates based filtered by resource

### PR DESCRIPTION
**Pull request by kanika1894**
For scenarios such as right click on cluster, the performance can be improved if instead of getting all the templates and then filtering it by resource type, we can get templates based on the resource type itself.

If we already hve a resource selected, we can pass the language and deploy target to get the rquired templates.

### CODE REVIEW

File: src/configure/helper/templateHelper.ts
Comment (anuragc617) [inline src/configure/helper/templateHelper.ts:172]: this code is repeated in below function as well, can you create separate function for it with name convertToRemotePipelineTemplate

File: src/configure/clients/ITemplateServiceClient.ts
Comment (anuragc617) [inline src/configure/clients/ITemplateServiceClient.ts:11]: function name should be GetTemplatesInfoByFilter and provide all three args like this
getTemplatesInfoByFilter( languageFilter: string, deployTargetFilter?: string, buildTargetFilter?: string) : Promise<TemplateInfo[]>

File: src/configure/configure.ts
Comment (kanika1894) [reply src/configure/configure.ts:176]: Isn't it the same thing?
Anyway we'll have to put up a condition on the resourceNode type itself to put the required values for deployTarget and language, so instead of ``` context['isResourceSelected] && resourceNode.type === TargetResourceType.AKS ```, the existing one looke better?

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:177]: This will break the azure pipelines flow. `getSelectedPipeline` handles both local templates (which I thin is used for azure pipelines case) and remote templates (which is used for gh workflow case). It decides the kind of template to use based on the repo provider and some extension settings.
Hence it would not be wise to bifurcate here. We need to go deeper inside getSelectedPipeline and then only call getFilteredTemplates in case of remote templates.

Comment (kanika1894) [reply src/configure/configure.ts:177]: ohh! yes! 
will make the required changes.

File: src/configure/clients/github/TemplateServiceClient.ts
Comment (bishal-pdMSFT) [inline src/configure/clients/github/TemplateServiceClient.ts:24]: this function should take language, build Target and deployTarget as parameters. The switch case needs to move to the caller

Comment (bishal-pdMSFT) [reply src/configure/clients/github/TemplateServiceClient.ts:24]: The reason is that the client class should be as thin as possible and very specific to constructs of the service it is trying to  call. VS Code business constructs should not be much visible here.

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:187]: it seems `resourceNode` is always going to be null as `analyzeNode()` always return null

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:188]: Do we need to explicitly throw this error? Anyways getTemplatesByRepoAnalysis will return empty array of templates and hence a n error will be shown to user

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:504]: I guess you have removed the check for `templateServiceEnabled`. But if you want to do so then you should do across the complete code and remove the variable definition as well. Otherwise it is not complete cleanup

File: src/configure/clients/github/TemplateServiceClient.ts
Comment (kanika1894) [reply src/configure/clients/github/TemplateServiceClient.ts:24]: cool, will update

File: src/configure/configure.ts
Comment (kanika1894) [reply src/configure/configure.ts:187]: line 268
```return this.inputs.targetResource.resource;```
It returns the resource in case of right click scenario.

File: src/configure/configure.ts
Comment (kanika1894) [reply src/configure/configure.ts:188]: If the repo has something besides AKS, it will cause an issue and the flow will try to continue with webApp template and will break. So, anyway, we know that this scenario is not supported by us. therefore, explictly throwing. Better to have a proper error message to use instead of some dev exception.

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [reply src/configure/configure.ts:187]: Ah I missed that addition

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:268]: It looks weird to me that the function return a value in case it is a resource node but otherwise returns null if it is a folder node.
We should keep it generic and not return anything in either case. We are anyways setting `this.context['isResourceAlreadySelected'] = true;` which should be good way to check later whether resource is already selected.

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [reply src/configure/configure.ts:187]: As mentioned in another comment, instead of using `resourceNode` variable, we should rely on a function called `isResourceAlreadySelected` which will internally check if  this.context['isResourceAlreadySelected'] is true

Comment (kanika1894) [reply src/configure/configure.ts:187]: yes, that we can do
Or anyway as we are initializing the targetResource as a part of inputs, that can be used as well.
Because for cases, where we require resource details when it is selected, we will have to use the ``` inputs.targetResource.resource ```

File: src/configure/configure.ts
Comment (kanika1894) [reply src/configure/configure.ts]: updated

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:171]: this should not be async

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [reply src/configure/configure.ts:506]: Did you handle this comment?

Comment (kanika1894) [reply src/configure/configure.ts:506]: yes, it was there only in model.ts. Have removed it.

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:163]: this should not be async

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [reply src/configure/configure.ts:192]: Okay, so after reading through and thinking a lot I think this is okay. I can justify this explicit condition so that we do not have to start filtering local templates by target resource type.

File: src/configure/configure.ts
Comment (bishal-pdMSFT) [inline src/configure/configure.ts:192]: I think you asked Pulkit for an error message. Is this final?

**Review by bishal-pdMSFT — APPROVED** [review]

File: src/configure/configure.ts
Comment (kanika1894) [reply src/configure/configure.ts:192]: Still waiting for his approval. Will merge only after I get that.

Comment (bishal-pdMSFT) [reply src/configure/configure.ts:192]: @kanika1894  Lets merge this PR so that you can do bugbash tomorrow. the string can anyways be changed easily later.

**Pull request merged**
